File: server_java/examples/src/main/java/io/grpc/examples/raft/RaftServer.java
package io.grpc.examples.raft;

import io.grpc.*;
import io.grpc.stub.StreamObserver;

import java.io.IOException;
import java.util.*;
import java.util.concurrent.*;
import java.util.logging.Logger;
import raft.RaftServiceGrpc;
import raft.RaftServiceOuterClass.*;

public class RaftServer {
    private static final Logger logger = Logger.getLogger(RaftServer.class.getName());

    private final int serverId;
    private final List<String> peers;
    private final int port;
    private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
    private Server server;

    private String state = "Follower";
    private int currentTerm = 0;
    private Integer votedFor = null;
    private Integer leaderId = null;
    private final List<LogEntry> logs = new ArrayList<>();
    private long lastHeartbeat = System.currentTimeMillis();
    private final Random random = new Random();

    public RaftServer(int serverId, int port, List<String> peers) {
        this.serverId = serverId;
        this.port = port;
        this.peers = peers;
    }

    private void start() throws IOException {
        server = Grpc.newServerBuilderForPort(port, InsecureServerCredentials.create())
                .addService(new RaftServiceImpl())
                .build()
                .start();
        logger.info("Server " + serverId + " started, listening on " + port);
        startElectionTimer();
    }

    private void startElectionTimer() {
        scheduler.scheduleAtFixedRate(() -> {
            if (System.currentTimeMillis() - lastHeartbeat > randomElectionTimeout()) {
                startElection();
            }
        }, 100, 100, TimeUnit.MILLISECONDS);
    }

    private void startHeartbeatTimer() {
        scheduler.scheduleAtFixedRate(this::sendHeartbeat, 0, 100, TimeUnit.MILLISECONDS);
    }

    private void startElection() {
        state = "Candidate";
        currentTerm++;
        votedFor = serverId;
        leaderId = null; // Reset leaderId during elections
        logger.info("Server " + serverId + " became Candidate for term " + currentTerm);

        int votes = 1; // Vote for itself
        for (String peer : peers) {
            boolean voteGranted = sendRequestVote(peer);
            if (voteGranted) {
                votes++;
            }
        }

        if (votes > peers.size() / 2) {
            state = "Leader";
            leaderId = serverId;
            logger.info("Server " + serverId + " became Leader for term " + currentTerm);
            startHeartbeatTimer();
        }
    }

    private boolean sendRequestVote(String peer) {
        ManagedChannel channel = ManagedChannelBuilder.forTarget(peer).usePlaintext().build();
        RaftServiceGrpc.RaftServiceBlockingStub stub = RaftServiceGrpc.newBlockingStub(channel);
        RequestVoteRequest request = RequestVoteRequest.newBuilder()
                .setTerm(currentTerm)
                .setCandidateId(serverId)
                .setLastLogIndex(logs.size() - 1)
                .setLastLogTerm(logs.isEmpty() ? 0 : logs.get(logs.size() - 1).getTerm())
                .build();
        try {
            RequestVoteResponse response = stub.requestVote(request);
            logger.info("Server " + serverId + " received RequestVoteResponse from peer: " + peer);
            return response.getVoteGranted();
        } catch (Exception e) {
            logger.warning("Failed to send RequestVote to peer " + peer + ": " + e.getMessage());
            return false;
        } finally {
            channel.shutdown();
        }
    }

    private void sendHeartbeat() {
        if (state.equals("Leader")) {
            logger.info("Server " + serverId + " sends heartbeat to peers");
            for (String peer : peers) {
                sendAppendEntries(peer);
            }
        }
    }

    private void sendAppendEntries(String peer) {
        ManagedChannel channel = ManagedChannelBuilder.forTarget(peer).usePlaintext().build();
        RaftServiceGrpc.RaftServiceBlockingStub stub = RaftServiceGrpc.newBlockingStub(channel);
        AppendEntriesRequest request = AppendEntriesRequest.newBuilder()
                .setTerm(currentTerm)
                .setLeaderId(serverId)
                .setPrevLogIndex(logs.size() - 1)
                .setPrevLogTerm(logs.isEmpty() ? 0 : logs.get(logs.size() - 1).getTerm())
                .setLeaderCommit(0)
                .build();
        try {
            AppendEntriesResponse response = stub.appendEntries(request);
            logger.info("Server " + serverId + " received AppendEntriesResponse from peer: " + peer);
        } catch (Exception e) {
            logger.warning("Failed to send AppendEntries to peer " + peer + ": " + e.getMessage());
        } finally {
            channel.shutdown();
        }
    }

    private int randomElectionTimeout() {
        return 150 + random.nextInt(150);
    }

    public static void main(String[] args) throws IOException, InterruptedException {
        if (args.length < 3) {
            System.err.println("Usage: RaftServer <serverId> <port> <peer1> <peer2> ...");
            System.exit(1);
        }
        int serverId = Integer.parseInt(args[0]);
        int port = Integer.parseInt(args[1]);
        List<String> peers = Arrays.asList(Arrays.copyOfRange(args, 2, args.length));

        final RaftServer raftServer = new RaftServer(serverId, port, peers);
        raftServer.start();
    }

    private class RaftServiceImpl extends RaftServiceGrpc.RaftServiceImplBase {
        @Override
        public void requestVote(RequestVoteRequest request, StreamObserver<RequestVoteResponse> responseObserver) {
            logger.info("Server " + serverId + " received RequestVote from " + request.getCandidateId());
            boolean voteGranted = false;
            if (request.getTerm() > currentTerm || (request.getTerm() == currentTerm
                    && (votedFor == null || votedFor == request.getCandidateId()))) {
                voteGranted = true;
                votedFor = request.getCandidateId();
                currentTerm = request.getTerm();
                lastHeartbeat = System.currentTimeMillis();
            }
            responseObserver
                    .onNext(RequestVoteResponse.newBuilder().setTerm(currentTerm).setVoteGranted(voteGranted).build());
            responseObserver.onCompleted();
        }

        @Override
        public void appendEntries(AppendEntriesRequest request,
                StreamObserver<AppendEntriesResponse> responseObserver) {
            logger.info("Server " + serverId + " received AppendEntries from " + request.getLeaderId());
            boolean success = false;
            if (request.getTerm() >= currentTerm) {
                currentTerm = request.getTerm();
                leaderId = request.getLeaderId();
                lastHeartbeat = System.currentTimeMillis();
                success = true;
            }
            responseObserver
                    .onNext(AppendEntriesResponse.newBuilder().setTerm(currentTerm).setSuccess(success).build());
            responseObserver.onCompleted();
        }

        @Override
        public void clientRequest(ClientRequestMessage request,
                StreamObserver<ClientResponseMessage> responseObserver) {
            logger.info("Server " + serverId + " received ClientRequest: " + request.getCommand());
            if (!state.equals("Leader")) {
                if (leaderId != null) {
                    String leaderAddress = "localhost:" + (5000 + leaderId);
                    logger.info("Server " + serverId + " forwarding request to leader at " + leaderAddress);
                    ManagedChannel channel = ManagedChannelBuilder.forTarget(leaderAddress).usePlaintext().build();
                    RaftServiceGrpc.RaftServiceBlockingStub stub = RaftServiceGrpc.newBlockingStub(channel);
                    try {
                        ClientResponseMessage response = stub.clientRequest(request);
                        responseObserver.onNext(response);
                    } catch (Exception e) {
                        logger.warning("Failed to forward request to leader: " + e.getMessage());
                        responseObserver.onNext(ClientResponseMessage.newBuilder().setSuccess(false)
                                .setMessage("Leader unreachable").build());
                    } finally {
                        channel.shutdown();
                    }
                } else {
                    responseObserver.onNext(ClientResponseMessage.newBuilder().setSuccess(false)
                            .setMessage("No leader available").build());
                }
            } else {
                logs.add(LogEntry.newBuilder().setIndex(logs.size()).setTerm(currentTerm)
                        .setCommand(request.getCommand()).build());
                responseObserver.onNext(
                        ClientResponseMessage.newBuilder().setSuccess(true).setMessage("Command applied").build());
            }
            responseObserver.onCompleted();
        }
    }
}

File: server_python/examples/python/raft/raft_server.py
import grpc
import time
import threading
import random
from concurrent import futures
from raft_service_pb2 import (
    RequestVoteRequest, 
    RequestVoteResponse, 
    AppendEntriesResponse, 
    AppendEntriesRequest,
    ClientRequestMessage,
    ClientResponseMessage,
    LogEntry as ProtoLogEntry,
    GetStatusResponse,
    GetLogResponse
)
import raft_service_pb2_grpc


class LogEntry:
    def __init__(self, index, term, command):
        self.index = index
        self.term = term
        self.command = command


class RaftServer(raft_service_pb2_grpc.RaftServiceServicer):
    def __init__(self, server_id, peers):
        self.server_id = server_id
        self.peers = peers  # List of other server addresses
        self.leader_id = None  # Initialize leader_id to track the current leader
        self.unreachable_peers = {}  # Track unreachable peers and their retry times
        self.retry_interval = 5  # Retry unreachable peers every 5 seconds
        self.current_term = 0
        self.voted_for = None
        self.state = "Follower"  # Follower, Candidate, Leader
        self.logs = []  # Log entries: list of LogEntry
        self.commit_index = -1  # Index of the last committed log entry
        self.next_index = {}  # Next log index to send to each peer (Leader state)
        self.match_index = {}  # Highest log index known to be replicated on each peer (Leader state)
        self.heartbeat_interval = 0.1  # 100ms
        self.election_timeout = random.uniform(0.15, 0.3)  # Random between 150ms and 300ms
        self.election_timer = None
        self.heartbeat_timer = None
        self.vote_count = 0  # To track votes received during an election
        self.reset_election_timer()

    def reset_election_timer(self):
        if self.election_timer:
            self.election_timer.cancel()
        self.election_timer = threading.Timer(self.election_timeout, self.start_election)
        self.election_timer.start()

    def start_heartbeat_timer(self):
        if self.heartbeat_timer:
            self.heartbeat_timer.cancel()
        self.heartbeat_timer = threading.Timer(self.heartbeat_interval, self.send_heartbeat)
        self.heartbeat_timer.start()

    def start_election(self):
        self.state = "Candidate"
        self.current_term += 1
        self.voted_for = self.server_id
        self.vote_count = 1  # Vote for itself
        print(f"Process {self.server_id} transitions to Candidate for term {self.current_term}")
        self.reset_election_timer()  # Reset election timer for a new term
        self.send_request_vote()

    def send_request_vote(self):
        print(f"Process {self.server_id} sends RPC RequestVote to peers")
        for peer in self.peers:
            threading.Thread(target=self.request_vote_from_peer, args=(peer,)).start()

    def request_vote_from_peer(self, peer):
        # Check if the peer is in the unreachable list and if its retry timer has passed
        if peer in self.unreachable_peers:
            if time.time() < self.unreachable_peers[peer]:
                print(f"Process {self.server_id}: Skipping unreachable peer {peer} for RequestVote.")
                return  # Skip if still within retry interval
            else:
                print(f"Process {self.server_id}: Retrying previously unreachable peer {peer} for RequestVote.")
        
        with grpc.insecure_channel(peer) as channel:
            stub = raft_service_pb2_grpc.RaftServiceStub(channel)
            request = RequestVoteRequest(
                term=self.current_term,
                candidateId=self.server_id,
                lastLogIndex=len(self.logs) - 1,
                lastLogTerm=self.logs[-1].term if self.logs else 0,
            )
            try:
                response = stub.RequestVote(request)
                print(f"Process {self.server_id} receives RPC RequestVoteResponse from Process {peer}: {response.voteGranted}")
                if response.voteGranted:
                    self.vote_count += 1
                    if self.vote_count > len(self.peers) // 2:
                        self.state = "Leader"
                        print(f"Process {self.server_id} transitions to Leader for term {self.current_term}")
                        self.initialize_leader_state()
                        self.start_heartbeat_timer()
                # Remove peer from unreachable list if it responds successfully.
                if peer in self.unreachable_peers:
                    del self.unreachable_peers[peer]
            except grpc.RpcError as e:
                print(f"RPC error while contacting Process {peer}: {e}")
                # Mark the peer as unreachable and set a new retry time
                self.unreachable_peers[peer] = time.time() + self.retry_interval

    def initialize_leader_state(self):
        self.leader_id = self.server_id  # Set leader_id to self
        for peer in self.peers:
            self.next_index[peer] = len(self.logs)
            self.match_index[peer] = -1
        self.send_heartbeat()  # Immediately send heartbeats after election

    def send_heartbeat(self):
        if self.state == "Leader":
            print(f"Process {self.server_id} sends AppendEntries (heartbeat) to peers")
            for peer in self.peers:
                threading.Thread(target=self.append_entries_to_peer, args=(peer,)).start()
            self.start_heartbeat_timer()  # Schedule next heartbeat
    
    # Get status method for test script.
    def GetStatus(self, request, context):
        """
        Handle GetStatus RPC to provide the server's current status.
        """
        print(f"Process {self.server_id} responding to GetStatus request.")
        return GetStatusResponse(
            state=self.state,
            current_term=self.current_term,
            voted_for=self.voted_for if self.voted_for is not None else -1,
            leader_id=self.leader_id if self.leader_id is not None else -1
        )
    
    def to_proto_log_entry(self, log_entry):
        """
        Convert internal LogEntry to ProtoLogEntry.
        """
        return ProtoLogEntry(
            index=log_entry.index,
            term=log_entry.term,
            command=log_entry.command
        )

    def GetLog(self, request, context):
        """
        Handle GetLog RPC to provide the server's log entries.
        """
        print(f"Process {self.server_id} responding to GetLog request.")
        proto_logs = [self.to_proto_log_entry(log) for log in self.logs]
        return GetLogResponse(entries=proto_logs)



    def append_entries_to_peer(self, peer):
        if peer in self.unreachable_peers:
            if time.time() < self.unreachable_peers[peer]:
                print(f"Process {self.server_id}: Skipping unreachable peer {peer} for AppendEntries.")
                return  # Skip if still within retry interval
            else:
                print(f"Process {self.server_id}: Retrying previously unreachable peer {peer} for AppendEntries.")
        
        with grpc.insecure_channel(peer) as channel:
            stub = raft_service_pb2_grpc.RaftServiceStub(channel)
            prev_log_index = self.next_index[peer] - 1
            prev_log_term = self.logs[prev_log_index].term if prev_log_index >= 0 else 0
            entries = [
                ProtoLogEntry(index=log.index, term=log.term, command=log.command)
                for log in self.logs[self.next_index[peer]:]
            ]
            request = AppendEntriesRequest(
                term=self.current_term,
                leaderId=self.server_id,
                entries=entries,
                prevLogIndex=prev_log_index,
                prevLogTerm=prev_log_term,
                leaderCommit=self.commit_index,
            )
            try:
                response = stub.AppendEntries(request)
                print(f"Process {self.server_id} receives AppendEntriesResponse from Process {peer}: {response.success}")
                if response.success:
                    self.match_index[peer] = len(self.logs) - 1
                    self.next_index[peer] = len(self.logs)
                else:
                    self.next_index[peer] -= 1  # Decrement and retry
                # Remove peer from unreachable list if it responds successfully
                if peer in self.unreachable_peers:
                    del self.unreachable_peers[peer]
            except grpc.RpcError as e:
                print(f"Process {self.server_id}: Failed to contact peer {peer} for AppendEntries. Error: {e.details()} - {e.code()}")
                # Mark the peer as unreachable and set a new retry time
                self.unreachable_peers[peer] = time.time() + self.retry_interval
    
    def get_leader_address(self):
        """
        Return the address of the leader based on self.leader_id.
        """
        if self.leader_id is not None:
            leader_port = 5000 + self.leader_id  # Assuming ports are sequential
            leader_address = f"localhost:{leader_port}"
            print(f"Process {self.server_id}: Leader ID {self.leader_id} mapped to {leader_address}")
            return leader_address
        print(f"Process {self.server_id}: No valid leader_id found")
        return None

    def ClientRequest(self, request, context):
        print(f"Process {self.server_id} received ClientRequest: {request.command}")
        print(f"Process {self.server_id}: State = {self.state}, Leader ID = {self.leader_id}, Term = {self.current_term}")
        
        if self.state != "Leader":
            # Forward to the leader
            leader_address = self.get_leader_address()
            if leader_address:
                print(f"Process {self.server_id}: Forwarding request to leader at {leader_address}")
                with grpc.insecure_channel(leader_address) as channel:
                    stub = raft_service_pb2_grpc.RaftServiceStub(channel)
                    forward_request = ClientRequestMessage(command=request.command)
                    try:
                        forward_response = stub.ClientRequest(forward_request)
                        return forward_response
                    except grpc.RpcError as e:
                        print(f"Process {self.server_id}: Failed to forward to leader {leader_address}. Error: {e.details()}")
                        return ClientResponseMessage(success=False, message="Leader unreachable")
            else:
                print(f"Process {self.server_id}: No leader available to forward request.")
                return ClientResponseMessage(success=False, message="No leader available")
        
        # Leader processes the command
        log_entry = LogEntry(len(self.logs), self.current_term, request.command)
        self.logs.append(log_entry)
        self.replicate_log()
        return ClientResponseMessage(success=True, message="Command applied")


    def replicate_log(self):
        print(f"Process {self.server_id} replicates log entries to followers")
        for peer in self.peers:
            threading.Thread(target=self.append_entries_to_peer, args=(peer,)).start()

    def RequestVote(self, request, context):
        print(f"Process {self.server_id} receives RPC RequestVote from Process {request.candidateId}")
        vote_granted = False
        if request.term > self.current_term:
            self.current_term = request.term
            self.state = "Follower"
            self.voted_for = None
        if (self.voted_for is None or self.voted_for == request.candidateId) and request.term >= self.current_term:
            self.voted_for = request.candidateId
            vote_granted = True
            self.reset_election_timer()
            print(f"Process {self.server_id} grants vote to Process {request.candidateId}: {vote_granted}")
        return RequestVoteResponse(term=self.current_term, voteGranted=vote_granted)

    def AppendEntries(self, request, context):
        print(f"Process {self.server_id} receives RPC AppendEntries from Process {request.leaderId}")
        print(f"Process {self.server_id}: Current Leader ID = {self.leader_id}, Received Leader ID = {request.leaderId}")
        success = False
        if request.term >= self.current_term:
            self.current_term = request.term
            self.state = "Follower"
            if self.leader_id != request.leaderId:  # Log leader_id updates
                self.leader_id = request.leaderId
                print(f"Process {self.server_id}: Updated leader_id to {self.leader_id}")
            self.reset_election_timer()
            # Validate log consistency
            if request.prevLogIndex == -1 or (
                len(self.logs) > request.prevLogIndex
                and self.logs[request.prevLogIndex].term == request.prevLogTerm
            ):
                success = True
                # Append new entries if applicable
                try:
                    self.logs = self.logs[:request.prevLogIndex + 1] + [
                        LogEntry(entry.index, entry.term, entry.command) for entry in request.entries
                    ]
                    # Update commit index
                    if request.leaderCommit > self.commit_index:
                        self.commit_index = min(request.leaderCommit, len(self.logs) - 1)
                except IndexError:
                    print(f"Process {self.server_id}: IndexError during log replication")
                    success = False
        print(f"Process {self.server_id}: AppendEntries success: {success}, Leader ID: {self.leader_id}")
        return AppendEntriesResponse(term=self.current_term, success=success)

def serve(server_id, port, peers):
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    raft_service_pb2_grpc.add_RaftServiceServicer_to_server(RaftServer(server_id, peers), server)
    server.add_insecure_port(f'[::]:{port}')
    print(f"Process {server_id} started, listening on {port}")
    server.start()
    try:
        server.wait_for_termination()
    except KeyboardInterrupt:
        print(f"Process {server_id} shutting down")
        server.stop(0)  # Gracefully stop the server

if __name__ == "__main__":
    import sys

    if len(sys.argv) < 3:
        print("Usage: python raft_server.py <server_id> <port> <peer1> <peer2> ...")
        sys.exit(1)

    server_id = int(sys.argv[1])
    port = int(sys.argv[2])
    peers = sys.argv[3:]
    serve(server_id, port, peers)